### PR TITLE
Refine and fix bugs about firmware update tool and watchdog issue

### PR DESCRIPTION
**Pull request by huaqianli**
iot2050-firmware-update: Fix bug in class destruction and refactor code
iot2050-module-firmware-update: Add module firmware update
iot2050-eio-manager: Refine the firmware update code
linux: Disable lock-step for all iot2050 boards
u-boot: Disable lock-step mode for all iot2050 boards
firmware-update-package: Add two more default uboot envs

### CODE REVIEW

File: recipes-kernel/linux/files/patches-6.1/0097--arm64-dts-ti-iot2050-Disable-lock-step-for-all-iot20.patch [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+partially lacks support for lock-step mode. This implies that all
+iot2050 boards can't support this mode. As a result, lock-step mode has
+been disabled across all iot2050 boards for consistency and to avoid
+potential issues.
+
+Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>
+---
+ arch/arm64/boot/dts/ti/k3-am65-iot2050-common-pg2.dtsi | 5 -----
+ arch/arm64/boot/dts/ti/k3-am65-iot2050-common.dtsi     | 5 +++++
+ arch/arm64/boot/dts/ti/k3-am6528-iot2050-basic.dts     | 5 -----
```
Comment (jan-kiszka) [inline recipes-kernel/linux/files/patches-6.1/0097--arm64-dts-ti-iot2050-Disable-lock-step-for-all-iot20.patch:17]: You forgot to clean up also k3-am6548-iot2050-advanced-m2.dts.

File: recipes-bsp/u-boot/files/0011-arm-dts-iot2050-Disable-lock-step-mode-for-all-iot20.patch [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+partially lacks support for lock-step mode. This implies that not all
+iot2050 boards can support this mode. As a result, lock-step mode has
+been disabled across all iot2050 boards for consistency and to avoid
+potential issues.
+
+Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>
+---
+ arch/arm/dts/k3-am65-iot2050-common-pg2.dtsi | 5 -----
+ arch/arm/dts/k3-am65-iot2050-common.dtsi     | 5 +++++
+ arch/arm/dts/k3-am6528-iot2050-basic.dts     | 5 -----
```
Comment (jan-kiszka) [inline recipes-bsp/u-boot/files/0011-arm-dts-iot2050-Disable-lock-step-mode-for-all-iot20.patch:17]: Same here: k3-am6548-iot2050-advanced-m2.dts

File: recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            print("Error: Failed to write firmware B")
+            return 1
+
+    print("Firmware updated successfully")
+
+    return 0
+
+if __name__ == '__main__':
+    CODE = main(sys.argv[1:])
+    sys.exit(CODE)
```
Comment (jan-kiszka) [inline recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl:84]: missing newline at EOF

File: recipes-app/iot2050-module-firmware-update/iot2050-module-firmware-update_01.bb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# Copyright (c) Siemens AG, 2024
+#
+# Authors:
+#  Li Hua Qian <huaqian.li@siemens.com>
+#
+# This file is subject to the terms and conditions of the MIT License.  See
+# COPYING.MIT file in the top-level directory.
+#
+
+DESCRIPTION = "Module Firmware Update Scripts"
```
Comment (jan-kiszka) [inline recipes-app/iot2050-module-firmware-update/iot2050-module-firmware-update_01.bb:11]: What is is "Module" in this context?

File: recipes-kernel/linux/files/patches-6.1/0097--arm64-dts-ti-iot2050-Disable-lock-step-for-all-iot20.patch (hunk 1 above)
Comment (huaqianli) [reply recipes-kernel/linux/files/patches-6.1/0097--arm64-dts-ti-iot2050-Disable-lock-step-for-all-iot20.patch:17]: It was already cleaned up in https://github.com/siemens/meta-iot2050/commit/34b7ef3c9af6a1abb3d36cf9c0a8822741577117.

File: recipes-bsp/u-boot/files/0011-arm-dts-iot2050-Disable-lock-step-mode-for-all-iot20.patch (hunk 2 above)
Comment (huaqianli) [reply recipes-bsp/u-boot/files/0011-arm-dts-iot2050-Disable-lock-step-mode-for-all-iot20.patch:17]: It doesn't exist in k3-am6548-iot2050-advanced-m2.dts since no patch to do it in u-boot. I propose it was cleaned up during preparing https://github.com/siemens/meta-iot2050/commit/bc65912f96051c5bdf1b298a280a7efb84d14dfa.

File: recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl (hunk 3 above)
Comment (huaqianli) [reply recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl]: ACK and done

File: recipes-app/iot2050-module-firmware-update/iot2050-module-firmware-update_01.bb (hunk 4 above)
Comment (huaqianli) [reply recipes-app/iot2050-module-firmware-update/iot2050-module-firmware-update_01.bb]: Refined

File: recipes-kernel/linux/files/patches-6.1/0097--arm64-dts-ti-iot2050-Disable-lock-step-for-all-iot20.patch (hunk 1 above)
Comment (jan-kiszka) [reply recipes-kernel/linux/files/patches-6.1/0097--arm64-dts-ti-iot2050-Disable-lock-step-for-all-iot20.patch:17]: Oh, indeed, was looking at stale version of the kernel here.

File: recipes-bsp/u-boot/files/0011-arm-dts-iot2050-Disable-lock-step-mode-for-all-iot20.patch (hunk 2 above)
Comment (jan-kiszka) [reply recipes-bsp/u-boot/files/0011-arm-dts-iot2050-Disable-lock-step-mode-for-all-iot20.patch:17]: It does - in upstream, but that is due to a missing sync from the Linux kernel. Not relevant in this context.

File: recipes-app/iot2050-module-firmware-update/iot2050-module-firmware-update_01.bb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+# Copyright (c) Siemens AG, 2024
+#
+# Authors:
+#  Li Hua Qian <huaqian.li@siemens.com>
+#
+# This file is subject to the terms and conditions of the MIT License.  See
+# COPYING.MIT file in the top-level directory.
+#
+
+DESCRIPTION = "Customized Module(Special DI) Firmware Update Scripts"
```
Comment (jan-kiszka) [inline recipes-app/iot2050-module-firmware-update/iot2050-module-firmware-update_01.bb:11]: At what is "Special DI" now?

File: recipes-core/images/iot2050-image-example.bb [hunk 6]
```diff
@@ -35,6 +35,7 @@ IMAGE_INSTALL += " \
     customizations-example \
     switchserialmode \
     iot2050-firmware-update \
+    iot2050-module-firmware-update \
```
Comment (jan-kiszka) [inline recipes-core/images/iot2050-image-example.bb:38]: I don't get the commit message in line with with addition of the "module" update package to the image.

File: recipes-app/iot2050-module-firmware-update/iot2050-module-firmware-update_01.bb (hunk 5 above)
Comment (BaochengSu) [reply recipes-app/iot2050-module-firmware-update/iot2050-module-firmware-update_01.bb:11]: @huaqianli: 

IOT2050 Customized external signal module, such as `SM SENS DI`. This kind of module is not PLC1200 standard, but "customized" by IOT2050. Hence such module could not be upgraded within the PLC ecosystem, but we have to provide method to support the FW upgrading for them.

File: recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl [hunk 7]
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3
```
Comment (jan-kiszka) [inline recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl:1]: First sentence of the commit message seems to lack a verb.

File: recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl [hunk 8]
```diff
@@ -142,22 +142,20 @@ class UpgradeError(Exception):
...
 
 class Firmware():
     """The Firmware class represents flash base operations for all flashes"""
     def __init__(self, firmware):
-        try:
-            self.firmware = open(firmware, "rb")
-        except (IOError, TypeError):
-            self.firmware = firmware
+        if not hasattr(firmware, 'read'):
+            raise UpgradeError("TypeError: firmware must be a file-like object!")
```
Comment (jan-kiszka) [inline recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl:152]: nit: `isinstance(firmware, io.IOBase)` would be more accurate.

File: recipes-app/iot2050-firmware-update/files/test_iot2050_firmware_update.py [hunk 9]
```diff
@@ -0,0 +1,800 @@
...
+        # Test with force=True
+        mock_args.force = True
+        mock_args.rollback = False
+        result = main([])
+        self.assertEqual(result, None)
+        mock_force_update.assert_called_once_with(mock_interactor, 'firmware.bin')
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (jan-kiszka) [inline recipes-app/iot2050-firmware-update/files/test_iot2050_firmware_update.py:800]: When and how will those unit tests be run?

File: recipes-devtools/firmware-update-package/files/update.conf.json.tmpl [hunk 10]
```diff
@@ -33,6 +33,8 @@
         }
     ],
     "suggest_preserved_uboot_env": [
-        "boot_targets"
+        "boot_targets",
+        "m2_manual_config",
+        "watchdog_timeout_ms"
```
Comment (BaochengSu) [inline recipes-devtools/firmware-update-package/files/update.conf.json.tmpl:38]: This should be in a separate commit.

File: recipes-devtools/firmware-update-package/files/iot2050-generate-fwu-tarball.sh [hunk 11]
```diff
@@ -9,16 +9,43 @@
...
-    sed -i '/"description": ".*"/s/V.*["$]/'$2\"'/g' $1
-    sed -i '/"min_version": ".*"/s/"V/"'$version_letter'/g' $1
+
+    sed -i '/"version": ".*"/s|"V.*"|"'$2'"|g' $1
+    sed -i '/"description": ".*"/s|V.*["$]|'$2\"'|g' $1
+    sed -i '/"min_version": ".*"/s|"V|"'$version_letter'|g' $1
+
+    # if the SUGGEST_PRESERVED_UBOOT_ENV is not empty, update the
+    # suggest_preserved_uboot_env in the update.conf.json
+    if [ ! -z "$SUGGEST_PRESERVED_UBOOT_ENV" ]; then
```
Comment (BaochengSu) [inline recipes-devtools/firmware-update-package/files/iot2050-generate-fwu-tarball.sh:21]: Add a new yml template, then pass the SCM value via this yaml.

File: recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    )
+    args = parser.parse_args(args)
+
+    if not args.firmware_a and not args.firmware_b:
+        print("Error: No firmware file specified")
+        parser.print_help()
+        return 1
+
+    slot = args.slot
+    slot_path = f"/eiofs/controller/slot{slot}"
```
Comment (BaochengSu) [inline recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl:50]: You don't have to define a `slot` for the very next sentence.

File: recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl [hunk 13]
```diff
@@ -356,6 +351,9 @@ class EnvFirmware(Firmware):
             if self.env_mtd_num and self.env_bk_mtd_num:
                 break
 
+            if mtd_num > 20:
+                raise UpgradeError("EnvFirmware: No env partition found")
+
```
Comment (BaochengSu) [inline recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl:356]: I would use `for ... else ...`, to make it more pythonic :)

Comment (BaochengSu) [reply recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl:356]: Also add a comment to explain 20 is enough for now.

Comment (BaochengSu) [reply recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl:356]: Or is it better to wrap this logic into mtd_device.get_partition_by_label("env")?

Comment (huaqianli) [reply recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl:356]: Done. However, given the development and testing effort involved, it's advisable to avoid making significant modifications to the original logic.

File: recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl (hunk 12 above)
Comment (huaqianli) [reply recipes-app/iot2050-module-firmware-update/files/iot2050-module-firmware-update.tmpl:50]: Done

File: recipes-devtools/firmware-update-package/files/iot2050-generate-fwu-tarball.sh (hunk 11 above)
Comment (huaqianli) [reply recipes-devtools/firmware-update-package/files/iot2050-generate-fwu-tarball.sh]: As we discussed, remove this undue feature.

File: recipes-devtools/firmware-update-package/files/update.conf.json.tmpl (hunk 10 above)
Comment (huaqianli) [reply recipes-devtools/firmware-update-package/files/update.conf.json.tmpl:38]: Done

File: recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl [hunk 14]
```diff
@@ -371,13 +369,13 @@ class EnvFirmware(Firmware):
...
-                            mtd_dev_path, mtd_size, mtd_erasesize,
-                            env_default_binary, firmware_size
-                        )
+                    firmware_size = self.mtd_device.write(
+                        mtd_dev_path, mtd_size, mtd_erasesize,
+                        env_default_binary, firmware_size
+                    )
+
+                    if firmware_size > 0:
+                        raise UpgradeError("Write env failed")
```
Comment (BaochengSu) [inline recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl:378]: Are you sure it can write all the bytes in one shot? I think the original logic is trying to avoid if the data could not be written in just one write.

Comment (huaqianli) [reply recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl:378]: If we keep the loop, there will be a problem; If somehow it always fails to write, a dead loop will happen here.

On the high level there are 3 times retries to improve this situation.

Comment (huaqianli) [reply recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl:378]: Also there is a loop in `mtd_device.write` to confirm all data are touched.

```py
            while mtd_pos < mtd_size and file_size > 0:
                mtd_content = os.read(mtd_dev, mtd_erasesize)
                firmware_content = file_obj.read(mtd_erasesize)
                padsize = mtd_erasesize - len(firmware_content)
                firmware_content += bytearray([0xff] * padsize)

                if not mtd_content == firmware_content:
                    self.__erase(mtd_dev, mtd_pos, mtd_erasesize)
                    os.lseek(mtd_dev, mtd_pos, os.SEEK_SET)
                    os.write(mtd_dev, firmware_content)

                mtd_pos += mtd_erasesize
                file_size -= mtd_erasesize
```

**Review by BaochengSu — APPROVED** [review]
